apis/spaces/v1alpha1/incontrolplaneoverride_types.go

```diff
@@ -0,0 +1,293 @@
+// Copyright 2024 Upbound Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"regexp"
+	"strings"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+)
+
+var (
+	regexFieldNotDeclared = regexp.MustCompile(`failed to create typed patch object.+field not declared in schema`)
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Namespaced,categories=spaces
+
+// InControlPlaneOverride represents resource configuration overrides in
+// a ControlPlane. The specified override can be applied on single objects
+// as well as claim/XR object hierarchies.
+type InControlPlaneOverride struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   InControlPlaneOverrideSpec   `json:"spec"`
+	Status InControlPlaneOverrideStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// InControlPlaneOverrideList is a list of InControlPlaneOverride objects.
+type InControlPlaneOverrideList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []InControlPlaneOverride `json:"items"`
+}
+
+// PropagationMode denotes the traversal direction on an object's hierarchy.
+type PropagationMode string
```

Comment:
I think we should prepend `Patch` to most (if not all) types here.
```suggestion
type PatchPropagationMode string
```

Reply:
Done.